### PR TITLE
nixos/firefox-syncserver: fix setup failure due to duplicate key

### PR DESCRIPTION
**Pull request by symphorien**
in some circumstances, the setup service would fail with: ERROR 1062 (23000) at line 5: Duplicate entry '1' for key 'PRIMARY' so let's use an upsert instead of insert. This also simplifies the script.

Besides, also fix that when the setup script changes, the corresponding script is not restarted as it is usually not active, so we trigger a restart of the main systemd service.

cc @pennae @mweinelt 
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by pennae — COMMENTED** [review]
in which circumstances does the script fail? ideally we'd want the setup script to run only once *at all*, on the first service start after `singleNode.enable` was set. any further changes to the database made by the user should not be reset blindly, those could well have been meaningful.

File: nixos/modules/services/networking/firefox-syncserver.nix [hunk 1]
```diff
@@ -32,6 +32,44 @@ let
...
+          mysql ${cfg.database.name} -Ne 'SHOW TABLES' | grep -q services
+        }
+
+        update_config() {
+          mysql ${cfg.database.name} <<"EOF"
+            BEGIN;
+
+            INSERT INTO `services` (`id`, `service`, `pattern`)
+              VALUES (1, 'sync-1.5', '{node}/1.5/{uid}')
+              ON DUPLICATE KEY UPDATE service='sync-1.5', pattern='{node}/1.5/{uid}';
```
[the remark this answers is not in the document]
Comment (symphorien) [reply nixos/modules/services/networking/firefox-syncserver.nix:49]: well that's exactly what I expect of a declarative configuration, tbh.

File: nixos/modules/services/networking/firefox-syncserver.nix [hunk 2]
```diff
@@ -32,6 +32,44 @@ let
...
+            INSERT INTO `nodes` (`id`, `service`, `node`, `available`, `current_load`,
+                                 `capacity`, `downed`, `backoff`)
+              VALUES (1, 1, '${cfg.singleNode.url}', ${toString cfg.singleNode.capacity},
+              0, ${toString cfg.singleNode.capacity}, 0, 0)
+              ON DUPLICATE KEY UPDATE node = '${cfg.singleNode.url}', available=${toString cfg.singleNode.capacity}, capacity=${toString cfg.singleNode.capacity};
```
[the remark this answers is not in the document]
Comment (symphorien) [reply nixos/modules/services/networking/firefox-syncserver.nix:54]: changed

**Comment by symphorien** [comment]
I setup the minimal configuration in the manual, the first time mariadb failed to start due to a timeout, then I rebuilt a few times to change the nginx part of the config and I got this error message: `ERROR 1062 (23000) at line 5: Duplicate entry '1' for key 'PRIMARY' `
Here are the relevant logs:
```
2022-10-30T13:53:26.509 169551712894 init.scope systemd[1] WARNING /etc/systemd/system/firefox-syncserver.service:41: Unknown key name 'Stderr' in section 'Service', ignoring.
2022-10-30T13:53:35.239 169560443211 init.scope systemd[1] INFO Starting MySQL Server...
2022-10-30T13:53:38.362 169563565484 mysql.service mysql-pre-start[1849188] INFO chown: changing ownership of '/nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/lib/mysql/plugin/auth_pam_tool_dir/auth_pam_tool': Read-only file system
2022-10-30T13:53:38.363 169563566469 mysql.service mysql-pre-start[1849106] INFO Couldn't set an owner to '/nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/lib/mysql/plugin/auth_pam_tool_dir/auth_pam_tool'.
2022-10-30T13:53:38.363 169563566469 mysql.service mysql-pre-start[1849106] INFO It must be root, the PAM authentication plugin doesn't work otherwise..
2022-10-30T13:53:38.367 169563571129 mysql.service mysql-pre-start[1849189] INFO chown: changing ownership of '/nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/lib/mysql/plugin/auth_pam_tool_dir': Read-only file system
2022-10-30T13:53:38.369 169563572544 mysql.service mysql-pre-start[1849106] INFO Cannot change ownership of the '/nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/lib/mysql/plugin/auth_pam_tool_dir' directory
2022-10-30T13:53:38.369 169563572793 mysql.service mysql-pre-start[1849106] INFO to the 'mysql' user. Check that you have the necessary permissions and try again.
2022-10-30T13:53:38.370 169563573556 mysql.service mysql-pre-start[1849106] INFO Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
2022-10-30T13:53:44.386 169569589601 mysql.service mysql-pre-start[1849194] INFO 2022-10-30 13:53:44 0 [Warning] WSREP: Failed to guess base node address. Set it explicitly via wsrep_node_address.
2022-10-30T13:53:44.386 169569589601 mysql.service mysql-pre-start[1849194] INFO 2022-10-30 13:53:44 0 [Warning] WSREP: Failed to guess base node address. Set it explicitly via wsrep_node_address.
2022-10-30T13:53:44.386 169569589601 mysql.service mysql-pre-start[1849194] INFO 2022-10-30 13:53:44 0 [Warning] WSREP: Guessing address for incoming client connections failed. Try setting wsrep_node_incoming_address explicitly.
2022-10-30T13:55:05.400 169650604113 init.scope systemd[1] WARNING mysql.service: start-pre operation timed out. Terminating.
2022-10-30T13:55:05.404 169650607995 init.scope systemd[1] WARNING mysql.service: Control process exited, code=killed, status=15/TERM
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO ERROR: 1105  Bootstrap file error, return code (0). Nearest query: 'CREATE DEFINER='mariadb.sys'@'localhost' PROCEDURE diagnostics (
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO         IN in_max_runtime int unsigned, IN in_interval int unsigned,
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO         IN in_auto_config enum ('current', 'medium', 'full')
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO     )
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO     COMMENT '
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO              Description
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO              Create a report of the current status of the server for diagnostics purposes. Data collected includes (some items depends on versions and settings):
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO                 * The GLOBAL VARIABLES
2022-10-30T13:55:09.598 169654802117 mysql.service mysql-pre-start[1849194] INFO                
2022-10-30T13:55:11.641 169656844975 init.scope systemd[1] WARNING mysql.service: Failed with result 'timeout'.
2022-10-30T13:55:11.645 169656849098 init.scope systemd[1] ERROR Failed to start MySQL Server.
2022-10-30T13:55:11.647 169656851277 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=mysql comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
2022-10-30T13:55:11.648 169656851922 - kernel[-] STATS audit: type=1130 audit(1667134511.646:34093): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=mysql comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
2022-10-30T13:55:11.648 169656852195 init.scope systemd[1] WARNING Dependency failed for firefox-syncserver.service.
2022-10-30T13:55:11.651 169656854761 init.scope systemd[1] WARNING Dependency failed for firefox-syncserver-setup.service.
2022-10-30T13:55:11.652 169656856314 init.scope systemd[1] STATS firefox-syncserver-setup.service: Job firefox-syncserver-setup.service/start failed with result 'dependency'.
2022-10-30T13:55:11.653 169656856837 init.scope systemd[1] STATS firefox-syncserver.service: Job firefox-syncserver.service/start failed with result 'dependency'.
2022-10-30T13:55:11.653 169656857069 init.scope systemd[1] INFO mysql.service: Consumed 2.967s CPU time, no IP traffic.
2022-10-30T13:59:03.360 169888564223 init.scope systemd[1] INFO Starting MySQL Server...
2022-10-30T13:59:03.700 169888903419 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Note] /nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/bin/mysqld (server 10.6.10-MariaDB) starting as process 1852255 ...
2022-10-30T13:59:03.720 169888923788 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Warning] WSREP: Failed to guess base node address. Set it explicitly via wsrep_node_address.
2022-10-30T13:59:03.720 169888923788 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Warning] WSREP: Failed to guess base node address. Set it explicitly via wsrep_node_address.
2022-10-30T13:59:03.720 169888923788 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Warning] WSREP: Guessing address for incoming client connections failed. Try setting wsrep_node_incoming_address explicitly.
2022-10-30T13:59:03.720 169888923788 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Note] WSREP: Node addr:
2022-10-30T13:59:03.885 169889088570 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:03 0 [Note] mysqld: Aria engine: starting recovery
2022-10-30T13:59:04.376 169889580124 mysql.service mysql-start[1852255] INFO recovered pages: 0% 10% 26% 36% 46% 59% 71% 81% 91% 100% (0.1 seconds); tables to flush: 5 4 3 2 1 0
2022-10-30T13:59:04.535 169889738655 mysql.service mysql-start[1852255] INFO  (0.4 seconds);
2022-10-30T13:59:04.535 169889738655 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] mysqld: Aria engine: recovery done
2022-10-30T13:59:04.560 169889764286 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Compressed tables use zlib 1.2.12
2022-10-30T13:59:04.566 169889770021 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Number of pools: 1
2022-10-30T13:59:04.566 169889770347 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
2022-10-30T13:59:04.567 169889771228 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Using liburing
2022-10-30T13:59:04.568 169889772260 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728
2022-10-30T13:59:04.571 169889774652 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Completed initialization of buffer pool
2022-10-30T13:59:04.613 169889816747 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Starting crash recovery from checkpoint LSN=13559,13559
2022-10-30T13:59:04.994 169890198164 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:04 0 [Note] InnoDB: Starting final batch to recover 18 pages from redo log.
2022-10-30T13:59:05.005 169890208434 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: 128 rollback segments are active.
2022-10-30T13:59:05.005 169890208989 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: Removed temporary tablespace data file: "./ibtmp1"
2022-10-30T13:59:05.005 169890208989 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2022-10-30T13:59:05.006 169890209404 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2022-10-30T13:59:05.006 169890209656 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2022-10-30T13:59:05.010 169890213881 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: 10.6.10 started; log sequence number 41958; transaction id 14
2022-10-30T13:59:05.011 169890214540 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] Plugin 'FEEDBACK' is disabled.
2022-10-30T13:59:05.011 169890214988 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
2022-10-30T13:59:05.012 169890215460 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] InnoDB: Cannot open '/var/lib/mysql/ib_buffer_pool' for reading: No such file or directory
2022-10-30T13:59:05.238 169890441917 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] Server socket created on IP: '0.0.0.0'.
2022-10-30T13:59:05.238 169890442198 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] Server socket created on IP: '::'.
2022-10-30T13:59:05.343 169890547150 mysql.service mysql-start[1852255] INFO 2022-10-30 13:59:05 0 [Note] /nix/store/c1lx0j75sjz3jgg2nncpygw7qhf6magg-mariadb-server-10.6.10/bin/mysqld: ready for connections.
2022-10-30T13:59:05.343 169890547150 mysql.service mysql-start[1852255] INFO Version: '10.6.10-MariaDB'  socket: '/run/mysqld/mysqld.sock'  port: 3306  MariaDB Server
2022-10-30T13:59:06.557 169891760399 init.scope systemd[1] INFO Started MySQL Server.
2022-10-30T13:59:06.559 169891763010 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=mysql comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T13:59:06.560 169891763731 - kernel[-] STATS audit: type=1130 audit(1667134746.555:34112): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=mysql comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:00:36.142 169981346247 init.scope systemd[1] INFO Started firefox-syncserver.service.
2022-10-30T14:00:36.144 169981348103 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:00:36.145 169981348762 - kernel[-] STATS audit: type=1130 audit(1667134836.143:34127): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:00:36.156 169981359442 - kernel[-] STATS audit: type=1130 audit(1667134836.152:34130): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:00:36.156 169981359821 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:00:36.156 169981360346 init.scope systemd[1] INFO Started firefox-syncserver-setup.service.
2022-10-30T14:01:19.619 170024822892 init.scope systemd[1] INFO firefox-syncserver-setup.service: Deactivated successfully.
2022-10-30T14:01:19.624 170024827775 - kernel[-] STATS audit: type=1131 audit(1667134879.621:34131): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:01:19.624 170024828225 - audit[1] INFO SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO Oct 30 13:14:36.758 ERRO Internal Server Error: TokenserverError { status: "internal-error", location: Internal, name: "", description: "A database error occurred: NotFound", http_status: 500, context: "A database error occurred: NotFound", backtrace:    0: <syncstorage_db_common::error::DbError as core::convert::From<diesel::result::Error>>::from
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    1: syncstorage::tokenserver::db::models::TokenserverDb::get_service_id_sync
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    2: <F as threadpool::FnBox>::call_box
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    3: std::sys_common::backtrace::__rust_begin_short_backtrace
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    4: core::ops::function::FnOnce::call_once{{vtable.shim}}
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    5: std::sys::unix::thread::Thread::new::thread_start
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    6: start_thread
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO    7: __GI___clone
2022-10-30T14:14:36.758 170821962342 firefox-syncserver.service syncstorage[1853573] INFO , token_type: Oauth }
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO Oct 30 13:22:52.346 ERRO Internal Server Error: TokenserverError { status: "internal-error", location: Internal, name: "", description: "A database error occurred: NotFound", http_status: 500, context: "A database error occurred: NotFound", backtrace:    0: <syncstorage_db_common::error::DbError as core::convert::From<diesel::result::Error>>::from
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    1: syncstorage::tokenserver::db::models::TokenserverDb::get_service_id_sync
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    2: <F as threadpool::FnBox>::call_box
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    3: std::sys_common::backtrace::__rust_begin_short_backtrace
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    4: core::ops::function::FnOnce::call_once{{vtable.shim}}
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    5: std::sys::unix::thread::Thread::new::thread_start
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    6: start_thread
2022-10-30T14:22:52.346 171317550243 firefox-syncserver.service syncstorage[1853573] INFO    7: __GI___clone
2022-10-30T14:22:52.348 171317552249 firefox-syncserver.service syncstorage[1853573] INFO , token_type: Oauth }
2022-10-30T14:27:23.179 171588382605 init.scope systemd[1] WARNING /etc/systemd/system/firefox-syncserver.service:41: Unknown key name 'Stderr' in section 'Service', ignoring.
2022-10-30T14:33:30.004 171955208067 init.scope systemd[1] INFO Stopping firefox-syncserver.service...
2022-10-30T14:33:31.925 171957128779 init.scope systemd[1] INFO firefox-syncserver.service: Deactivated successfully.
2022-10-30T14:33:31.931 171957134578 init.scope systemd[1] INFO Stopped firefox-syncserver.service.
2022-10-30T14:33:31.933 171957137002 - audit[1] INFO SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:33:31.934 171957137743 - kernel[-] STATS audit: type=1131 audit(1667136811.932:34798): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:33:31.936 171957140100 init.scope systemd[1] INFO firefox-syncserver.service: Consumed 2.630s CPU time, received 19.9K IP traffic, sent 8.7K IP traffic.
2022-10-30T14:33:37.743 171962946900 init.scope systemd[1] WARNING /etc/systemd/system/firefox-syncserver.service:41: Unknown key name 'Stderr' in section 'Service', ignoring.
2022-10-30T14:33:43.077 171968280802 init.scope systemd[1] INFO Started firefox-syncserver.service.
2022-10-30T14:33:43.086 171968289855 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:33:43.124 171968327702 init.scope systemd[1] INFO Started firefox-syncserver-setup.service.
2022-10-30T14:33:43.130 171968333638 - audit[1] INFO SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
2022-10-30T14:33:45.110 171970314258 firefox-syncserver-setup.service firefox-syncserver-setup-start[1906929] INFO ERROR 1062 (23000) at line 5: Duplicate entry '1' for key 'PRIMARY'
2022-10-30T14:33:45.113 171970316847 init.scope systemd[1] STATS firefox-syncserver-setup.service: Main process exited, code=exited, status=1/FAILURE
2022-10-30T14:33:45.114 171970317954 init.scope systemd[1] WARNING firefox-syncserver-setup.service: Failed with result 'exit-code'.
2022-10-30T14:33:45.121 171970324930 - audit[1] INFO SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=firefox-syncserver-setup comm="systemd" exe="/nix/store/77c9qlw0mmzbhl5f72sxqrxzrkyynds0-systemd-251.5/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
2022-10-30T14:34:16.967 172002171110 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:34:16.967 INFO Starting 2 workers
2022-10-30T14:34:16.978 172002181420 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:34:16.975 INFO Starting "actix-web-service-127.0.0.1:5000" service on 127.0.0.1:5000
2022-10-30T14:34:16.979 172002183000 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:34:16.979 INFO Server running on http://127.0.0.1:5000 (mysql) No quota
2022-10-30T14:35:37.027 172082230842 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:35:37.027 INFO {"uri.path":"/","ua.os.family":"Other","uri.method":"GET","ua.browser.family":"Other"}
2022-10-30T14:35:38.001 172083205185 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:35:38.001 INFO {"ua.browser.family":"Other","uri.method":"GET","ua.os.family":"Other","uri.path":"/"}
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:36:16.033 ERRO Internal Server Error: TokenserverError { status: "internal-error", location: Internal, name: "", description: "A database error occurred: NotFound", http_status: 500, context: "A database error occurred: NotFound", backtrace:    0: <syncstorage_db_common::error::DbError as core::convert::From<diesel::result::Error>>::from
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    1: syncstorage::tokenserver::db::models::TokenserverDb::get_service_id_sync
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    2: <F as threadpool::FnBox>::call_box
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    3: std::sys_common::backtrace::__rust_begin_short_backtrace
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    4: core::ops::function::FnOnce::call_once{{vtable.shim}}
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    5: std::sys::unix::thread::Thread::new::thread_start
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    6: start_thread
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO    7: __GI___clone
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO , token_type: Oauth }
2022-10-30T14:36:16.033 172121236991 firefox-syncserver.service syncstorage[1906888] INFO Oct 30 13:36:16.033 INFO {"ua.browser.family":"Firefox","uid":"foo","uri.method":"GET","metrics_uid":"bar","ua.os.family":"Linux","ua.os.ver":"UNKNOWN","ua.browser.ver":"106.0","ua":"106.0","uri.path":"/1.0/sync/1.5","ua.name":"Firefox"}
```
I had no mariadb before so no weird state to cause this.


About not changing manual changes by the user, I expect the singleNode option to be a sign-off by the user to declarative configuration, so resetting changes when the option is set is pretty much what I want.

File: nixos/modules/services/networking/firefox-syncserver.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (symphorien) [reply nixos/modules/services/networking/firefox-syncserver.nix:49]: being declarative and only supporting the easy case are two independent things. Making the single node case properly declarative does not prevent us from generalizing later to several nodes, nor for interested users to still use this module for a declarative+several nodes use cases. So this approach looks more robust to me (it solves an issue I experienced) and still flexible (people with complex needs can disable singleNode and do whatever) while respecting the intuition that a setting in my configuration is enforced declaratively.

**Pull request merged**